### PR TITLE
non-local means: support 3d+multichannel, 4d and 4d+multichannel

### LINKED ISSUE #5483: Allow use of complex numbers with denoise_nl_means
The code of non_local_means.py  converts the dataset to 'float', but I have a dataset with complex numbers. Is there a possibility to support more data types? 

Thanks
```

### PR DESCRIPTION
**Pull request by grlee77**
## Description

closes #5483 (3D complex-valued data can be processed as 2-channel 3D data)

This PR adds support for additional dimensions to non-local means. I had originally implemented this a couple of years ago for a research project and had never made a PR for it.  The recent request in #5483 lead to reviving this work.

The new dimensions are supported only for `fast_mode=True` as the slower mode becomes increasingly impractical in higher dimensions.

Although there are quite a few lines of code, the changes are straightforward extensions of the existing approach:
- The existing `_fast_nl_means_denoising_3d` has a new "channels" loop added.
- The new `_fast_nl_means_denoising_4d` is basically just a copy of `_fast_nl_means_denoising_3d`, but with one additional loop. 
- The new `_integral_to_distance_4d` follows an n-dimensional formula for this given in the reference by E. Tapia in that function's docstring

## Checklist

- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

File: skimage/restoration/tests/test_denoise.py [hunk 1]
```diff
@@ -515,13 +516,92 @@ def test_denoise_nl_means_multichannel(fast_mode, dtype, channel_axis):
...
+    img = np.zeros((8, 8, 8, 4, 4))
+    img[2:-2, 2:-2, 2:-2, 1:-1, :] = 1.
+    sigma = 0.3
+    imgn = img + sigma * np.random.randn(*img.shape)
+
+    psnr_noisy = peak_signal_noise_ratio(img, imgn, data_range=1.)
+
+    denoised_4dmc = restoration.denoise_nl_means(imgn, 3, 3, h=0.35 * sigma,
+                                                 fast_mode=True,
+                                                 multichannel=True,
```
Comment (jni) [inline skimage/restoration/tests/test_denoise.py:570]: channel_axis=-1?

**Review by jni — APPROVED** [review]
@grlee77 this is great! I haven't looked at the code in detail but tests look good and code does not raise red flags 😅. But I see a lot of `multichannel=<bool>` calls in the tests rather than `channel_axis=<int>`?

File: skimage/restoration/tests/test_denoise.py (hunk 1 above)
Comment (grlee77) [reply skimage/restoration/tests/test_denoise.py]: Yes, I missed updating these when doing the rebase! It should be fixed in the latest commit

File: skimage/restoration/_nl_means_denoising.pyx [hunk 2]
```diff
@@ -365,6 +366,53 @@ cdef inline double _integral_to_distance_3d(double[:, :, ::] integral,
...
+    """
+    References
+    ----------
+    J. Darbon, A. Cunha, T.F. Chan, S. Osher, and G.J. Jensen, Fast
+    nonlocal filtering applied to electron cryomicroscopy, in 5th IEEE
+    International Symposium on Biomedical Imaging: From Nano to Macro,
+    2008, pp. 1331-1334.
+
+    Jacques Froment. Parameter-Free Fast Pixelwise Non-Local Means
+    Denoising. Image Processing On Line, 2014, vol. 4, p. 300-326.
```
Comment (mkcor) [inline skimage/restoration/_nl_means_denoising.pyx:383]:
```suggestion
    Denoising. Image Processing On Line, 2014, vol. 4, pp. 300-326.
```
(just so it's consistent with the other references)

File: skimage/restoration/_nl_means_denoising.pyx [hunk 3]
```diff
@@ -365,6 +366,53 @@ cdef inline double _integral_to_distance_3d(double[:, :, ::] integral,
...
+
+    Used in _fast_nl_means_denoising_4d
+
+    The coefficients for the terms were determined using Eq. 54 of [1]_ as
+    implemented in the integral_image_coefficients function.
+
+    E. Tapia.  A note on the computation of high-dimensional integral images.
```
Comment (mkcor) [inline skimage/restoration/_nl_means_denoising.pyx:390]:
```suggestion
    .. [1] E. Tapia. A note on the computation of high-dimensional integral images.
```
?

File: skimage/restoration/_nl_means_denoising.pyx [hunk 4]
```diff
@@ -365,6 +366,53 @@ cdef inline double _integral_to_distance_3d(double[:, :, ::] integral,
...
+    Pattern Recognition Letters 2011. Vol. 32, pp.197-201.
```
Comment (mkcor) [inline skimage/restoration/_nl_means_denoising.pyx:391]:
```suggestion
    Pattern Recognition Letters, 2011, vol. 32, pp. 197-201.
```

File: skimage/restoration/_nl_means_denoising.pyx [hunk 5]
```diff
@@ -365,6 +366,53 @@ cdef inline double _integral_to_distance_3d(double[:, :, ::] integral,
...
+        integral[time - offset, pln + offset, row + offset, col + offset] -
+        integral[time + offset, pln - offset, row - offset, col - offset] +
+        integral[time + offset, pln - offset, row - offset, col + offset] +
+        integral[time + offset, pln - offset, row + offset, col - offset] -
+        integral[time + offset, pln - offset, row + offset, col + offset] +
+        integral[time + offset, pln + offset, row - offset, col - offset] -
+        integral[time + offset, pln + offset, row - offset, col + offset] -
+        integral[time + offset, pln + offset, row + offset, col - offset] +
+        integral[time + offset, pln + offset, row + offset, col + offset])
+    return  max(distance, 0.0) / (s4_h_square)
```
Comment (mkcor) [inline skimage/restoration/_nl_means_denoising.pyx:413]:
```suggestion
    return  max(distance, 0.0) / s4_h_square
```
not sure: was it supposed to help readability?

File: skimage/restoration/_nl_means_denoising.pyx [hunk 6]
```diff
@@ -483,6 +536,91 @@ cdef inline void _integral_image_3d(double [:, :, ::] padded,
...
+    t_col : Py_ssize_t
+        Shift along the column axis.
+    n_time : Py_ssize_t
+    n_pln : Py_ssize_t
+    n_row : Py_ssize_t
+    n_col : Py_ssize_t
+    var_diff : double
+        The double of the expected noise variance. If non-zero, this
+        is used to reduce the apparent patch distances by the expected
+         distance due to the noise.
```
Comment (mkcor) [inline skimage/restoration/_nl_means_denoising.pyx:571]:
```suggestion
        distance due to the noise.
```

File: skimage/restoration/_nl_means_denoising.pyx [hunk 7]
```diff
@@ -483,6 +536,91 @@ cdef inline void _integral_image_3d(double [:, :, ::] padded,
...
 
 
+cdef inline void _integral_image_4d(double [:, :, :, :, ::] padded,
+                                    double [:, :, :, ::] integral,
+                                    Py_ssize_t t_time, Py_ssize_t t_pln, Py_ssize_t t_row,
+                                    Py_ssize_t t_col, Py_ssize_t n_time, Py_ssize_t n_pln,
+                                    Py_ssize_t n_row, Py_ssize_t n_col, Py_ssize_t n_channels,
+                                    double var_diff) nogil:
+    """
+    Computes the integral of the squared difference between an image ``padded``
```
Comment (mkcor) [inline skimage/restoration/_nl_means_denoising.pyx:546]:
```suggestion
    """Compute the integral of the squared difference between an image ``padded``
```

File: skimage/restoration/_nl_means_denoising.pyx [hunk 8]
```diff
@@ -725,22 +872,196 @@ def _fast_nl_means_denoising_3d(cnp.ndarray[np_floats, ndim=3] image,
...
+    Parameters
+    ----------
+    image : ndarray
+        4-D input data to be denoised.
+    s : Py_ssize_t, optional
+        Size of patches used for denoising.
+    d : tuple of Py_ssize_t, optional
+        Maximal distance in pixels along each axis to search for patches used for denoising.
+    h : double, optional
+        cut-off distance (in gray levels). The higher h, the more permissive
```
Comment (mkcor) [inline skimage/restoration/_nl_means_denoising.pyx:917]:
```suggestion
        Cut-off distance (in gray levels). The higher h, the more permissive
```

File: skimage/restoration/_nl_means_denoising.pyx [hunk 9]
```diff
@@ -725,22 +872,196 @@ def _fast_nl_means_denoising_3d(cnp.ndarray[np_floats, ndim=3] image,
...
+
+    References
+    ----------
+    J. Darbon, A. Cunha, T.F. Chan, S. Osher, and G.J. Jensen, Fast
+    nonlocal filtering applied to electron cryomicroscopy, in 5th IEEE
+    International Symposium on Biomedical Imaging: From Nano to Macro,
+    2008, pp. 1331-1334.
+
+    Jacques Froment. Parameter-Free Fast Pixelwise Non-Local Means
+    Denoising. Image Processing On Line, 2014, vol. 4, p. 300-326.
```
Comment (mkcor) [inline skimage/restoration/_nl_means_denoising.pyx:936]:
```suggestion
    Denoising. Image Processing On Line, 2014, vol. 4, pp. 300-326.
```

**Review by mkcor — APPROVED** [review]
Great! I only followed the logic of the code by comparing it with existing `_fast_nl_means_denoising_2d` and `_fast_nl_means_denoising_3d`.

File: skimage/restoration/_nl_means_denoising.pyx [hunk 10]
```diff
@@ -725,22 +872,196 @@ def _fast_nl_means_denoising_3d(cnp.ndarray[np_floats, ndim=3] image,
...
+    cdef double [:, :, :, :, ::1] result = np.zeros_like(padded)
+    cdef double [:, :, :, ::1] weights = np.zeros_like(padded[..., 0])
+    cdef double [:, :, :, ::1] integral = np.zeros_like(padded[..., 0])
+    cdef Py_ssize_t n_pln, n_row, n_col, t_pln, t_row, t_col, \
+             pln, row, col, channel, n_channels, t_time, n_time, time
+    cdef Py_ssize_t time_dist_min, time_dist_max, pln_dist_min, pln_dist_max, \
+             row_dist_min, row_dist_max, col_dist_min, col_dist_max,
+    cdef Py_ssize_t d_row, d_col, d_pln, d_time
+    cdef double weight, distance, alpha
+    n_time, n_pln, n_row, n_col, n_channels = padded.shape[0], padded.shape[1], padded.shape[2], padded.shape[3], padded.shape[4]
```
Comment (rfezzani) [inline skimage/restoration/_nl_means_denoising.pyx:970]:
```suggestion
    n_time, n_pln, n_row, n_col, n_channels = padded.shape
```
Does tuple unpacking work here?

**Review by rfezzani — APPROVED** [review]
Thank you @grlee77. Just left a minor comment.

File: skimage/restoration/_nl_means_denoising.pyx (hunk 10 above)
Comment (grlee77) [reply skimage/restoration/_nl_means_denoising.pyx:970]: Apparently not!:
```
    Error compiling Cython file:
    ------------------------------------------------------------
    ...
                 pln, row, col, channel, n_channels, t_time, n_time, time
        cdef Py_ssize_t time_dist_min, time_dist_max, pln_dist_min, pln_dist_max, \
                 row_dist_min, row_dist_max, col_dist_min, col_dist_max,
        cdef Py_ssize_t d_row, d_col, d_pln, d_time
        cdef double weight, distance, alpha
        n_time, n_pln, n_row, n_col, n_channels = padded.shape
                                                       ^
    ------------------------------------------------------------

    skimage/restoration/_nl_means_denoising.pyx:970:52: Assignment to/from slice of wrong length, expected 8, got 5
```


elsewhere where image is a `cnp.ndarray` I get a different error instead if I use this style:
```
        cdef Py_ssize_t n_pln, n_row, n_col
        n_pln, n_row, n_col = image.shape
                                  ^
    ------------------------------------------------------------

    skimage/restoration/_nl_means_denoising.pyx:251:31: Cannot convert 'npy_intp *' to Python object
```

File: skimage/restoration/_nl_means_denoising.pyx (hunk 5 above)
Comment (grlee77) [reply skimage/restoration/_nl_means_denoising.pyx:413]: I think maybe this was leftover from a previous refactoring. Agree that it should just be removed

File: skimage/restoration/_nl_means_denoising.pyx (hunk 3 above)
Comment (grlee77) [reply skimage/restoration/_nl_means_denoising.pyx]: I have now fixed these across all functions where they are present in d1f0b28

**Comment by rfezzani** [comment]
I think @grlee77 answered all our concerns :smile:, after 3 approval, I think that we can merge this PR, thank you again @grlee77 :tada:

**Pull request merged**